lib/chef/resource/macos_userdefaults.rb

```diff
@@ -73,54 +111,96 @@ def coerce_booleans(val)
       end
 
       load_current_value do |desired|
-        value = coerce_booleans(desired.value)
-        cmd = "defaults read '#{desired.domain}' "
-        cmd << "'#{desired.key}' " if desired.key
-        cmd << " | grep -qx '#{value}'"
-
-        vc = if desired.user.nil?
-               shell_out(cmd)
-             else
-               shell_out(cmd, user: desired.user)
-             end
-
-        is_set !vc.error?
+        Chef::Log.debug "#load_current_value: shelling out \"#{defaults_export_cmd(desired).join(" ")}\" to determine state"
+        state = if desired.user.nil?
+                  shell_out(defaults_export_cmd(desired))
+                else
+                  shell_out(defaults_export_cmd(desired), user: desired.user)
+                end
+
+        if state.error? || state.stdout.empty?
+          Chef::Log.debug "#load_current_value: #{defaults_export_cmd(desired).join(" ")} returned stdout: #{state.stdout} and stderr: #{state.stderr}"
+          current_value_does_not_exist!
+        end
+
+        value ::Plist.parse_xml(state.stdout)
+      end
+
+      #
+      # The defaults command to export a domain
+      #
+      # @return [Array] defaults command
+      #
+      def defaults_export_cmd(resource)
+        state_cmd = ["/usr/bin/defaults"]
+
+        if resource.host == "current"
+          state_cmd << "-currentHost"
+        elsif resource.host # they specified a non-nil value, which is a hostname
+          state_cmd << ["-host", resource.host]
```

Comment:
```suggestion
          state_cmd += ["-host", resource.host]
```

Reply:
Isn't += is going to create a new object each time I add to the array vs. flatten doing it once?

Reply:
I switched to concat which avoids the object and avoids the flatten

Reply:
`+=` afaik, is the same thing as `concat` - it's the operator that calls `concat` underneath. Just like `<<` called `append`.

Reply:
```
irb(main):039:0> foo = []
irb(main):040:0> foo.object_id
=> 220
irb(main):041:0> foo =+ 'something'
irb(main):042:0> foo.object_id
=> 240
```

```
irb(main):044:0> foo = []
irb(main):045:0> foo.object_id
=> 260
irb(main):046:0> foo.concat(['bar'])
=> ["bar"]
irb(main):047:0> foo.object_id
=> 260
```

Reply:
That... code doesn't work:
```
irb#1(main):008:0> foo = []
irb#1(main):009:0> foo += 'stuff'
Traceback (most recent call last):
        2: from (irb#1):9
        1: from (irb#1):9:in `+'
TypeError (no implicit conversion of String into Array)
irb#1(main):010:0> 
```
But you're right, if you `+=` an array it makes a new object, that's strange:
```
irb#1(main):017:0> foo.object_id
=> 13540
irb#1(main):018:0> foo += ['stuff']
irb#1(main):019:0> foo.object_id
=> 13560
```

Is that more expensive than flatten?

Either way, I like `concat` as a solution. Best of both worlds.